File: modules/core/src/main/java/org/apache/ignite/binary/BinaryRawWriter.java
/*
 * Copyright 2019 GridGain Systems, Inc. and Contributors.
 *
 * Licensed under the GridGain Community Edition License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ignite.binary;

import java.math.BigDecimal;
import java.sql.Time;
import java.sql.Timestamp;
import java.util.Collection;
import java.util.Date;
import java.util.Map;
import java.util.UUID;

/**
 * Raw writer for binary object. Raw writer does not write field name hash codes, therefore,
 * making the format even more compact. However, if the raw writer is used,
 * dynamic structure changes to the binary objects are not supported.
 */
public interface BinaryRawWriter {
    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeByte(byte val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeShort(short val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeInt(int val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeLong(long val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeFloat(float val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeDouble(double val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeChar(char val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeBoolean(boolean val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeDecimal(BigDecimal val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeString(String val) throws BinaryObjectException;

    /**
     * @param val UUID to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeUuid(UUID val) throws BinaryObjectException;

    /**
     * @param val Date to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeDate(Date val) throws BinaryObjectException;

    /**
     * @param val Timestamp to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeTimestamp(Timestamp val) throws BinaryObjectException;

    /**
     * @param val Time to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeTime(Time val) throws BinaryObjectException;

    /**
     * @param obj Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeObject(Object obj) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeByteArray(byte[] val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeShortArray(short[] val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeIntArray(int[] val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeLongArray(long[] val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeFloatArray(float[] val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeDoubleArray(double[] val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeCharArray(char[] val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeBooleanArray(boolean[] val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeDecimalArray(BigDecimal[] val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeStringArray(String[] val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeUuidArray(UUID[] val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeDateArray(Date[] val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeTimestampArray(Timestamp[] val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeTimeArray(Time[] val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public void writeObjectArray(Object[] val) throws BinaryObjectException;

    /**
     * @param col Collection to write.
     * @throws BinaryObjectException In case of error.
     */
    public <T> void writeCollection(Collection<T> col) throws BinaryObjectException;

    /**
     * @param map Map to write.
     * @throws BinaryObjectException In case of error.
     */
    public <K, V> void writeMap(Map<K, V> map) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public <T extends Enum<?>> void writeEnum(T val) throws BinaryObjectException;

    /**
     * @param val Value to write.
     * @throws BinaryObjectException In case of error.
     */
    public <T extends Enum<?>> void writeEnumArray(T[] val) throws BinaryObjectException;
}

File: modules/platforms/cpp/binary/src/binary/binary_raw_writer.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "ignite/impl/binary/binary_writer_impl.h"
#include "ignite/binary/binary_raw_writer.h"

using namespace ignite::impl::binary;

namespace ignite
{
    namespace binary
    {
        BinaryRawWriter::BinaryRawWriter(BinaryWriterImpl* impl) : impl(impl)
        {
            // No-op.
        }

        void BinaryRawWriter::WriteInt8(int8_t val)
        {
            impl->WriteInt8(val);
        }

        void BinaryRawWriter::WriteInt8Array(const int8_t* val, int32_t len)
        {
            impl->WriteInt8Array(val, len);
        }

        void BinaryRawWriter::WriteBool(bool val)
        {
            impl->WriteBool(val);
        }

        void BinaryRawWriter::WriteBoolArray(const bool* val, int32_t len)
        {            
            impl->WriteBoolArray(val, len);
        }

        void BinaryRawWriter::WriteInt16(int16_t val)
        {
            impl->WriteInt16(val);
        }

        void BinaryRawWriter::WriteInt16Array(const int16_t* val, int32_t len)
        {
            impl->WriteInt16Array(val, len);
        }

        void BinaryRawWriter::WriteUInt16(uint16_t val)
        {
            impl->WriteUInt16(val);
        }

        void BinaryRawWriter::WriteUInt16Array(const uint16_t* val, int32_t len)
        {
            impl->WriteUInt16Array(val, len);
        }

        void BinaryRawWriter::WriteInt32(int32_t val)
        {
            impl->WriteInt32(val);
        }

        void BinaryRawWriter::WriteInt32Array(const int32_t* val, int32_t len)
        {
            impl->WriteInt32Array(val, len);
        }

        void BinaryRawWriter::WriteInt64(int64_t val)
        {
            impl->WriteInt64(val);
        }

        void BinaryRawWriter::WriteInt64Array(const int64_t* val, int32_t len)
        {
            impl->WriteInt64Array(val, len);
        }

        void BinaryRawWriter::WriteFloat(float val)
        {
            impl->WriteFloat(val);
        }

        void BinaryRawWriter::WriteFloatArray(const float* val, int32_t len)
        {
            impl->WriteFloatArray(val, len);
        }

        void BinaryRawWriter::WriteDouble(double val)
        {
            impl->WriteDouble(val);
        }

        void BinaryRawWriter::WriteDoubleArray(const double* val, int32_t len)
        {
            impl->WriteDoubleArray(val, len);
        }

        void BinaryRawWriter::WriteGuid(const Guid& val)
        {
            impl->WriteGuid(val);
        }

        void BinaryRawWriter::WriteGuidArray(const Guid* val, int32_t len)
        {
            impl->WriteGuidArray(val, len);
        }

        void BinaryRawWriter::WriteDate(const Date& val)
        {
            impl->WriteDate(val);
        }

        void BinaryRawWriter::WriteDateArray(const Date* val, int32_t len)
        {
            impl->WriteDateArray(val, len);
        }

        void BinaryRawWriter::WriteTimestamp(const Timestamp& val)
        {
            impl->WriteTimestamp(val);
        }

        void BinaryRawWriter::WriteTimestampArray(const Timestamp* val, int32_t len)
        {
            impl->WriteTimestampArray(val, len);
        }

        void BinaryRawWriter::WriteTime(const Time& val)
        {
            impl->WriteTime(val);
        }

        void BinaryRawWriter::WriteTimeArray(const Time* val, const int32_t len)
        {
            impl->WriteTimeArray(val, len);
        }

        void BinaryRawWriter::WriteString(const char* val)
        {
            if (val)
                WriteString(val, static_cast<int32_t>(strlen(val)));
            else
                WriteNull();
        }

        void BinaryRawWriter::WriteString(const char* val, int32_t len)
        {
            impl->WriteString(val, len);
        }

        BinaryStringArrayWriter BinaryRawWriter::WriteStringArray()
        {
            int32_t id = impl->WriteStringArray();

            return BinaryStringArrayWriter(impl, id);
        }

        void BinaryRawWriter::WriteBinaryEnum(BinaryEnumEntry entry)
        {
            impl->WriteBinaryEnum(entry);
        }

        void BinaryRawWriter::WriteNull()
        {
            impl->WriteNull();
        }
    }
}
